x/gov/abci.go

```diff
@@ -127,11 +127,11 @@ func EndBlocker(ctx sdk.Context, keeper *keeper.Keeper) error {
 			return false, err
 		}
 
-		// If an expedited proposal fails, we do not want to update
+		// If an expedited/optimistic proposal fails and isn't spammy, we do not want to update
 		// the deposit at this point since the proposal is converted to regular.
 		// As a result, the deposits are either deleted or refunded in all cases
-		// EXCEPT when an expedited proposal fails.
-		if passes || !(proposal.ProposalType == v1.ProposalType_PROPOSAL_TYPE_EXPEDITED) {
+		// EXCEPT when an expedited or optimistic proposal fails and isn't spammy.
+		if passes || (burnDeposits || !(proposal.ProposalType == v1.ProposalType_PROPOSAL_TYPE_EXPEDITED || proposal.ProposalType == v1.ProposalType_PROPOSAL_TYPE_OPTIMISTIC)) {
```

Comment:
This logic looks overly complicated, let's split it down a little to make it more readable.

Maybe something like this?

```go
if burnDeposits {
    err = keeper.DeleteAndBurnDeposits(ctx, proposal.Id)
} else if passes || (proposal.ProposalType != v1.ProposalType_PROPOSAL_TYPE_EXPEDITED && proposal.ProposalType != v1.ProposalType_PROPOSAL_TYPE_OPTIMISTIC) {
    err = keeper.RefundAndDeleteDeposits(ctx, proposal.Id)
}
```

Reply:
Good point, will do!